src/Expect.elm

```diff
@@ -546,84 +530,90 @@ reportFailure comparison expected actual =
         |> String.join "\n"
 
 
-reportCollectionFailure : String -> a -> b -> List c -> List d -> String
-reportCollectionFailure comparison expected actual missingKeys extraKeys =
-    [ toString actual
-    , "diff:" ++ formatDiffs Missing missingKeys ++ formatDiffs Extra extraKeys
-    , "╷"
-    , "│ " ++ comparison
-    , "╵"
-    , "diff:" ++ formatDiffs Extra missingKeys ++ formatDiffs Missing extraKeys
-    , toString expected
-    ]
-        |> String.join "\n"
-
-
-type Diff
-    = Extra
-    | Missing
-
+getFailureMessage : { given : String, description : String, reason : Test.Expectation.Reason } -> String
+getFailureMessage { given, description, reason } =
+    case reason of
+        Test.Expectation.Custom ->
+            description
+
+        Test.Expectation.Equals e a ->
+            verticalBar description e a
+
+        Test.Expectation.Comparison e a ->
+            verticalBar description e a
+
+        Test.Expectation.ListDiff e a ( i, itemE, itemA ) ->
+            verticalBar description e a
+                ++ "\n\nThe first diff is at index index "
+                ++ toString i
+                ++ ": it was `"
+                ++ itemA
+                ++ "`, but `"
+                ++ itemE
+                ++ "` was expected."
```

Comment:
🎨  worth considering moving out to a function. Also use `String.join ""` with a list for clearer syntax